### PR TITLE
ODH-ADR-Operator-0006-internal-api: Improve ODH Operator modularity

### PR DESCRIPTION
**Pull request by lburgazzoli**
## Description

ADR to describe the introduction of ODH Operator internal APIs to improve operator modularity and scalability.

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] ~Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).~
- [ ] ~The developer has manually tested the changes and verified that the changes work~

### CODE REVIEW

File: architecture-decision-records/operator/ODH-ADR-Operator-0006-internal-api.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+Move from a monolithic reconciliation loop to a more modular approach that manages components individually by:
+
+- Introducing more granular reconciliation.
+- Introducing a set of internal APIs/CRDs. 
+  - These CRDs will be used exclusively by the operator for internal management and will not be exposed to end-users. 
+  - Since CRDs are public, the ODH Operator will exclusively own them and it will revert any changes applied outside its control.
+- Each internal API/CRD will have its own reconciliation loop, which:
+  - Is responsible for managing the lifecycle and state of the associated components.
+  - Surfaces any relevant information as part of the component specific status (versions, namespace, routes, service, etc). 
+- The DSC reconciler would be in charge to create/delete internal CR depending on the management state of the components.
```
[the remark this answers is not in the document]
Comment (lburgazzoli) [reply architecture-decision-records/operator/ODH-ADR-Operator-0006-internal-api.md:58]: That's correct

File: architecture-decision-records/operator/ODH-ADR-Operator-0006-internal-api.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+* Provide additional, user facing CRDs. 
+
+## How
+
+Move from a monolithic reconciliation loop to a more modular approach that manages components individually by:
+
+- Introducing more granular reconciliation.
+- Introducing a set of internal APIs/CRDs. 
+  - These CRDs will be used exclusively by the operator for internal management and will not be exposed to end-users.
```
[the remark this answers is not in the document]
Comment (lburgazzoli) [reply architecture-decision-records/operator/ODH-ADR-Operator-0006-internal-api.md:53]: That's correct too

File: architecture-decision-records/operator/ODH-ADR-Operator-0006-internal-api.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jwforres) [reply architecture-decision-records/operator/ODH-ADR-Operator-0006-internal-api.md:58]: Can you add something into your bullets that makes step 5 clear? That is the bit I think was missing from the existing text.

[the remark this answers is not in the document]
Comment (lburgazzoli) [reply architecture-decision-records/operator/ODH-ADR-Operator-0006-internal-api.md]: Added a small adjustment, let me know if you want me to expand it

**Comment by jwforres** [comment]
This LGTM but giving the others a chance to review since this has impact to all components.

**Comment by lphiri** [comment]
@lburgazzoli  may I suggest adding the diagrams you had in the original proposal document? I found those really helpful to understand the "before" and "after".

**Comment by lburgazzoli** [comment]
> @lburgazzoli may I suggest adding the diagrams you had in the original proposal document? I found those really helpful to understand the "before" and "after".

done

**Review by etirelli — APPROVED** [review]
lgtm

**Review by jwforres — APPROVED** [review]

**Review by danielezonca — APPROVED** [review]
lgtm, small comment

File: architecture-decision-records/operator/ODH-ADR-Operator-0006-internal-api.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+This document outlines design decision to introduce additional, internal only, components specific CRDs.
+
+## Why
+
+The deployment of components is handled by a single reconcile loop that is in charge to deploy all the enabled components within the platform. 
+This means that one centralized loop handles the configuration, updates, and error handling for all components.
+
+However, this design introduced following drawbacks:
+- **Scalability**: as the ODH Operator evolves and new features or components are added, the single reconcile loop can become a significant bottleneck. The centralized nature of the loop means that any update or change in one component triggers a reconciliation for all components, even those unaffected by the change. This can lead to delays in processing, reduced responsiveness, and a potential decrease in the overall performance of the operator.
```
[the remark this answers is not in the document]
Comment (lburgazzoli) [reply architecture-decision-records/operator/ODH-ADR-Operator-0006-internal-api.md:35]: I would expect the CPU usage to decrease as the reconciliation would become more fine grained and impacting only components that have changed, so less resources spent rendering not impacted resources and less call to the API server.

**Pull request merged**